src/lib/xwalk-hack/churn.ts

```diff
@@ -0,0 +1,205 @@
+import logging = require('../logging/logging');
+import peerconnection = require('../webrtc/peerconnection');
+import random = require('random-lib');
+import signals = require('../webrtc/signals');
+
+import Candidate = candidate.Candidate;
+
+import ChurnSignallingMessage = churn_types.ChurnSignallingMessage;
+import TransformerConfig = churn_types.TransformerConfig;
+
+var log :logging.Log = new logging.Log('xchurn');
+
+export var getMidFromSdp = (sdp:string) : string => {
+  var mids :string[] = [];
+  sdp.split('\n').forEach((s) => {
+    var match = s.match(/^a=mid:(\S+)\s*$/);
+    if (match) {
+      mids.push(match[1]);
+    }
+  });
+  if (mids.length !== 1) {
+    throw new Error('Expected 1 mid, but there are actually ' + mids.length);
+  }
+  return mids[0];
+}
+
+// Generates a key suitable for use with CaesarCipher, viz. 1-255.
+var generateCaesarConfig_ = (): caesar.Config => {
+  return {
+    key: random.randomInt({
+      min: 1,
+      max: 255
+    })
+  };
+}
+
+/**
+ * A CHURN implementation for hacked crosswalk browsers.
+ * Caesar only.  Sets caesar parameter by SDP munging.
+ */
+export class Connection implements peerconnection.PeerConnection<ChurnSignallingMessage> {
+
+  // Number of instances created, for logging purposes.
+  private static id_ = 0;
+
+  // The "mid" 
+  private mid_ :string;
+
+  public peerOpenedChannelQueue :handler.QueueHandler<peerconnection.DataChannel, void>;
+  public signalForPeerQueue :handler.Queue<ChurnSignallingMessage, void>;
+
+  public onceConnected :Promise<void>;
+  public onceClosed :Promise<void>;
+
+  private pc_ :peerconnection.PeerConnectionClass;
+
+  // Fulfills once we know the obfuscator config, which may
+  // happen in response to a signalling channel message.
+  private haveTransformerConfig_ :(config:TransformerConfig) => void;
+  private onceHaveTransformerConfig_ = new Promise((F, R) => {
+    this.haveTransformerConfig_ = F;
+  });
+
+  constructor(rtcPc:freedom.RTCPeerConnection.RTCPeerConnection,
+              private name_ = 'unnamed-churn-' + Connection.id_,
+              private skipPublicEndpoint_?:boolean,
+              private portControl_?:freedom.PortControl.PortControl,
+              private preferredTransformerConfig_?:TransformerConfig) {
+    Connection.id_++;
+
+    this.signalForPeerQueue = new handler.Queue<ChurnSignallingMessage,void>();
+    this.configureConnection_(rtcPc);
+
+    // Forward onceXxx promises.
+    this.onceConnected = this.pc_.onceConnected;
+    this.onceClosed = this.pc_.onceClosed;
+  }
+
+  private configureConnection_ = (
+      freedomPc:freedom.RTCPeerConnection.RTCPeerConnection) => {
+    this.pc_ = new peerconnection.PeerConnectionClass(
+        freedomPc, this.name_);
+
+    // This will insert the caesar shift key
+    this.pc_.mungeLocalDescription = this.mungeLocalSdp_;
+
+    this.peerOpenedChannelQueue = this.pc_.peerOpenedChannelQueue;
+
+    this.pc_.signalForPeerQueue.setSyncHandler(
```

Comment:
I think mobile users are much less likely to be on a port control-ed network. I just realised that regular peerconnections don't do port control **even though there's no reason not to** - suggest we file an issue to make port control work for all `PeerConnection`s rather than repeating this again.